### ISSUE TITLE
[FEATURE] HHVM Support

### ISSUE THREAD

**Issue by cozylife** [bukka/php-fann#16]
I've noticed that this project appears to be inactive. However, I'm hoping that someone will create a branch that works with HHVM. I haven't actually tried it with HHVM yet, though I've noticed that the [docs say that FANN isn't _supported_ in HHVM](http://docs.hhvm.com/manual/en/fann.setup.php). If someone knows if it's incompatible or just that the team doesn't support FANN, I'd love to know.

At some point I'll try installing it on HHVM to play with, and I'll report my findings if I don't hear differently.

**Comment by bukka** [comment]
The project is still active. I actually plan to add support for PHP 7 later this year. There are no plans for support HHVM however. It would be probably completely new extension as the API is not compatible. Maybe there is an easier way (think that they have some API wrapper...) but I'm busy with other stuff and I won't have any time to write or even review/test chages for HHVM so it won't happen in this repo. This repo is just for PHP though.

**Comment by cozylife** [comment]
Is it incompatible just because of the Zend engine based functions or is it also incompatible for other reasons?

**Comment by bukka** [comment]
I don't know much about HHVM API so not sure. I guess you could start with this http://docs.hhvm.com/manual/en/configuration.zend.compat.php if you want to port it and then you will see.

However there will be some significant changes in the code base later when I port it to PHP 7 ( I'm going to use my [phpc layer](https://github.com/bukka/phpc/blob/master/phpc.h) after some additions ) so it might require further work then.

**Comment by cozylife** [comment]
[fann.c](https://github.com/bukka/php-fann/blob/master/fann.c) has a few things that need to be changed to work with the Zend compat. I searched all of the main code (not the tests), and I found that the following changes need to be made for it to work with [Zend compat](https://github.com/facebook/hhvm/tree/master/hphp/runtime/ext_zend_compat). I'm not a C programmer, so I don't know how to make the changes properly without lots of research.

* Use Z_RESVAL instead of Z_LVAL for resource access
* Allocate hashtables with ALLOC_HASHTABLE() and zvals with ALLOC_ZVAL() or one of the macros that calls ALLOC_ZVAL(), don't allocate them directly. Don't use malloc(sizeof(zval)) or create them on the stack.

You'll also need to create a stub file named ext_fann.php and fill it with a class structured like that of an interface. (By that I mean the functions are just stubs.) They use the hack syntax.

```
<<__NativeData("ZendCompat")>> class Foo {
  <<__Native("ZendCompat")>> function bar(mixed $a) : mixed;
}
```
The only things to notice are the ZendCompat stuff, and that every parameter uses the type "mixed" with a return type of "mixed." (: mixed is the return type) Don't worry about what each C function returns or the types of the parameters, as it doesn't matter in this case.

There are bugs in the Zend Compat that impact the way you design the compatibility. I've only listed the ones pertanant to FANN.

* Object destructors are currently not called.
* The object handlers (read_property, clone_obj, etc.) are not called either. (fann_connection.c and php_fann.h)
* Persistent HashTables and zvals will not work. (Maybe - Check fann_connection.c and fann.c)

If you're not interested in porting the code to Zend Compat or standard HHVM, perhaps someone else will find this useful in creating a branch.
